File: snippets/algorithms/graph-traversal/graph_traversal.py
#!/usr/bin/env python3
"""
Graph Traversal Algorithms Implementation

This module provides implementations of BFS and DFS for graph traversal,
along with a visualization function to display the traversal order.

Example usage:
    python graph_traversal_example.py
"""

from collections import deque
import time


class Graph:
    """A simple graph representation using adjacency lists."""
    
    def __init__(self):
        """Initialize an empty graph."""
        self.graph = {}
        
    def add_vertex(self, vertex):
        """Add a vertex to the graph if it doesn't exist."""
        if vertex not in self.graph:
            self.graph[vertex] = []
            
    def add_edge(self, v1, v2):
        """Add an edge between v1 and v2."""
        if v1 not in self.graph:
            self.add_vertex(v1)
        if v2 not in self.graph:
            self.add_vertex(v2)
            
        self.graph[v1].append(v2)
        self.graph[v2].append(v1)  # For undirected graph
        
    def bfs(self, start):
        """
        Perform Breadth-First Search starting from the given vertex.
        
        Args:
            start: The starting vertex for BFS
            
        Returns:
            A list containing vertices in the order they were visited
        """
        if start not in self.graph:
            return []
            
        visited = set([start])
        queue = deque([start])
        result = []
        
        print(f"Starting BFS traversal from vertex {start}")
        
        while queue:
            vertex = queue.popleft()
            result.append(vertex)
            
            print(f"Visiting: {vertex}")
            print(f"Queue: {list(queue)}")
            print(f"Visited so far: {result}")
            print("-" * 30)
            time.sleep(0.5)  # Slow down for demonstration
            
            for neighbor in sorted(self.graph[vertex]):  # Sort for consistent order
                if neighbor not in visited:
                    visited.add(neighbor)
                    queue.append(neighbor)
                    
        return result
        
    def dfs_recursive(self, start):
        """
        Perform Depth-First Search recursively starting from the given vertex.
        
        Args:
            start: The starting vertex for DFS
            
        Returns:
            A list containing vertices in the order they were visited
        """
        if start not in self.graph:
            return []
            
        visited = set()
        result = []
        
        print(f"Starting recursive DFS traversal from vertex {start}")
        
        def dfs_helper(vertex):
            visited.add(vertex)
            result.append(vertex)
            
            print(f"Visiting: {vertex}")
            print(f"Visited so far: {result}")
            print("-" * 30)
            time.sleep(0.5)  # Slow down for demonstration
            
            for neighbor in sorted(self.graph[vertex]):  # Sort for consistent order
                if neighbor not in visited:
                    dfs_helper(neighbor)
        
        dfs_helper(start)
        return result
        
    def dfs_iterative(self, start):
        """
        Perform Depth-First Search iteratively starting from the given vertex.
        
        Args:
            start: The starting vertex for DFS
            
        Returns:
            A list containing vertices in the order they were visited
        """
        if start not in self.graph:
            return []
            
        visited = set()
        stack = [start]
        result = []
        
        print(f"Starting iterative DFS traversal from vertex {start}")
        
        while stack:
            vertex = stack.pop()
            
            if vertex not in visited:
                visited.add(vertex)
                result.append(vertex)
                
                print(f"Visiting: {vertex}")
                print(f"Stack: {stack}")
                print(f"Visited so far: {result}")
                print("-" * 30)
                time.sleep(0.5)  # Slow down for demonstration
                
                # Add neighbors in reverse sorted order to simulate recursive DFS
                for neighbor in sorted(self.graph[vertex], reverse=True):
                    if neighbor not in visited:
                        stack.append(neighbor)
                        
        return result
    
    def visualize_graph(self):
        """Print a simple visualization of the graph structure."""
        print("\nGraph Structure:")
        print("-" * 30)
        for vertex, neighbors in sorted(self.graph.items()):
            print(f"{vertex} -> {sorted(neighbors)}")
        print("-" * 30)


def create_sample_graph():
    """Create a sample graph for demonstration."""
    g = Graph()
    
    # Add edges to build this graph:
    #     A
    #    / \
    #   B   C
    #  / \   \
    # D   E---F
    
    edges = [
        ('A', 'B'), ('A', 'C'),
        ('B', 'D'), ('B', 'E'),
        ('C', 'F'), ('E', 'F')
    ]
    
    for v1, v2 in edges:
        g.add_edge(v1, v2)
        
    return g


def main():
    """Main function to demonstrate graph traversal algorithms."""
    g = create_sample_graph()
    g.visualize_graph()
    
    print("\n=== BFS Traversal ===")
    bfs_result = g.bfs('A')
    print(f"BFS Result: {bfs_result}")
    
    print("\n=== DFS Traversal (Recursive) ===")
    dfs_rec_result = g.dfs_recursive('A')
    print(f"DFS Recursive Result: {dfs_rec_result}")
    
    print("\n=== DFS Traversal (Iterative) ===")
    dfs_iter_result = g.dfs_iterative('A')
    print(f"DFS Iterative Result: {dfs_iter_result}")


if __name__ == "__main__":
    main() 

File: snippets/algorithms/graph-traversal/graphTraversal.js
/**
 * Graph Traversal Algorithms in JavaScript
 * 
 * This module provides implementations of BFS and DFS for graph traversal.
 * 
 * Example usage:
 *   node graphTraversal.js
 */

class Graph {
  /**
   * Initialize an empty graph
   */
  constructor() {
    this.adjacencyList = {};
  }

  /**
   * Add a vertex to the graph
   * @param {string|number} vertex - The vertex to add
   */
  addVertex(vertex) {
    if (!this.adjacencyList[vertex]) {
      this.adjacencyList[vertex] = [];
    }
  }

  /**
   * Add an edge between two vertices
   * @param {string|number} v1 - First vertex
   * @param {string|number} v2 - Second vertex
   */
  addEdge(v1, v2) {
    if (!this.adjacencyList[v1]) {
      this.addVertex(v1);
    }
    if (!this.adjacencyList[v2]) {
      this.addVertex(v2);
    }
    
    this.adjacencyList[v1].push(v2);
    this.adjacencyList[v2].push(v1); // For undirected graph
  }

  /**
   * Breadth-First Search traversal
   * @param {string|number} start - Starting vertex
   * @returns {Array} - Vertices in the order they were visited
   */
  bfs(start) {
    if (!this.adjacencyList[start]) return [];
    
    const queue = [start];
    const visited = { [start]: true };
    const result = [];
    
    console.log(`Starting BFS traversal from vertex ${start}`);
    
    while (queue.length) {
      const currentVertex = queue.shift();
      result.push(currentVertex);
      
      console.log(`Visiting: ${currentVertex}`);
      console.log(`Queue: [${queue.join(', ')}]`);
      console.log(`Visited so far: [${result.join(', ')}]`);
      console.log('-'.repeat(30));
      
      // Get neighbors and sort for consistent order
      const neighbors = [...this.adjacencyList[currentVertex]].sort();
      
      for (const neighbor of neighbors) {
        if (!visited[neighbor]) {
          visited[neighbor] = true;
          queue.push(neighbor);
        }
      }
    }
    
    return result;
  }

  /**
   * Depth-First Search traversal (recursive)
   * @param {string|number} start - Starting vertex
   * @returns {Array} - Vertices in the order they were visited
   */
  dfsRecursive(start) {
    if (!this.adjacencyList[start]) return [];
    
    const visited = {};
    const result = [];
    
    console.log(`Starting recursive DFS traversal from vertex ${start}`);
    
    const dfs = (vertex) => {
      visited[vertex] = true;
      result.push(vertex);
      
      console.log(`Visiting: ${vertex}`);
      console.log(`Visited so far: [${result.join(', ')}]`);
      console.log('-'.repeat(30));
      
      // Get neighbors and sort for consistent order
      const neighbors = [...this.adjacencyList[vertex]].sort();
      
      for (const neighbor of neighbors) {
        if (!visited[neighbor]) {
          dfs(neighbor);
        }
      }
    };
    
    dfs(start);
    return result;
  }

  /**
   * Depth-First Search traversal (iterative)
   * @param {string|number} start - Starting vertex
   * @returns {Array} - Vertices in the order they were visited
   */
  dfsIterative(start) {
    if (!this.adjacencyList[start]) return [];
    
    const stack = [start];
    const visited = {};
    const result = [];
    
    console.log(`Starting iterative DFS traversal from vertex ${start}`);
    
    while (stack.length) {
      const currentVertex = stack.pop();
      
      if (!visited[currentVertex]) {
        visited[currentVertex] = true;
        result.push(currentVertex);
        
        console.log(`Visiting: ${currentVertex}`);
        console.log(`Stack: [${stack.join(', ')}]`);
        console.log(`Visited so far: [${result.join(', ')}]`);
        console.log('-'.repeat(30));
        
        // Get neighbors and sort in reverse order for stack
        const neighbors = [...this.adjacencyList[currentVertex]].sort().reverse();
        
        for (const neighbor of neighbors) {
          if (!visited[neighbor]) {
            stack.push(neighbor);
          }
        }
      }
    }
    
    return result;
  }

  /**
   * Print a visualization of the graph structure
   */
  visualizeGraph() {
    console.log('\nGraph Structure:');
    console.log('-'.repeat(30));
    
    // Sort the vertices for consistent output
    const vertices = Object.keys(this.adjacencyList).sort();
    
    for (const vertex of vertices) {
      const neighbors = [...this.adjacencyList[vertex]].sort();
      console.log(`${vertex} -> [${neighbors.join(', ')}]`);
    }
    
    console.log('-'.repeat(30));
  }
}

/**
 * Create a sample graph for demonstration
 * @returns {Graph} - A sample graph
 */
function createSampleGraph() {
  const g = new Graph();
  
  // Add edges to build this graph:
  //     A
  //    / \
  //   B   C
  //  / \   \
  // D   E---F
  
  const edges = [
    ['A', 'B'], ['A', 'C'],
    ['B', 'D'], ['B', 'E'],
    ['C', 'F'], ['E', 'F']
  ];
  
  for (const [v1, v2] of edges) {
    g.addEdge(v1, v2);
  }
  
  return g;
}

/**
 * Main function to demonstrate graph traversal algorithms
 */
function main() {
  const g = createSampleGraph();
  g.visualizeGraph();
  
  console.log('\n=== BFS Traversal ===');
  const bfsResult = g.bfs('A');
  console.log(`BFS Result: [${bfsResult.join(', ')}]`);
  
  console.log('\n=== DFS Traversal (Recursive) ===');
  const dfsRecResult = g.dfsRecursive('A');
  console.log(`DFS Recursive Result: [${dfsRecResult.join(', ')}]`);
  
  console.log('\n=== DFS Traversal (Iterative) ===');
  const dfsIterResult = g.dfsIterative('A');
  console.log(`DFS Iterative Result: [${dfsIterResult.join(', ')}]`);
}

// Run the main function if this file is executed directly
if (require.main === module) {
  main();
}

// Export the Graph class for use in other modules
module.exports = { Graph, createSampleGraph }; 